components/serverless/internal/controllers/serverless/validation.go

```diff
@@ -7,46 +7,84 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
 	"strings"
 )
 
 var _ stateFn = stateFnValidateFunction
 
+type validationFn func() []string
+
 func stateFnValidateFunction(_ context.Context, r *reconciler, s *systemState) (stateFn, error) {
 	rc := s.instance.Spec.ResourceConfiguration
 	fnResourceCfg := r.cfg.fn.ResourceConfig.Function.Resources
-	vrFunction := validateFunctionResources(rc, fnResourceCfg.MinRequestedCPU.Quantity, fnResourceCfg.MinRequestedMemory.Quantity)
+	validateFunctionResources := validateFunctionResourcesFn(rc, fnResourceCfg.MinRequestedCPU.Quantity, fnResourceCfg.MinRequestedMemory.Quantity)
 	buildResourceCfg := r.cfg.fn.ResourceConfig.BuildJob.Resources
-	vrBuild := validateBuildResources(rc, buildResourceCfg.MinRequestedCPU.Quantity, buildResourceCfg.MinRequestedMemory.Quantity)
+	validateBuildResources := validateBuildResourcesFn(rc, buildResourceCfg.MinRequestedCPU.Quantity, buildResourceCfg.MinRequestedMemory.Quantity)
+
+	validationFns := []validationFn{
+		validateFunctionResources,
+		validateBuildResources,
+		validateEnvs(s.instance.Spec.Env, "spec.env"),
+	}
+	validationResults := []string{}
+	for _, validationFn := range validationFns {
+		result := validationFn()
+		validationResults = append(validationResults, result...)
+	}
 
-	vr := append(vrFunction, vrBuild...)
-	if len(vr) != 0 {
-		msg := strings.Join(vr, ". ")
+	if len(validationResults) != 0 {
+		msg := strings.Join(validationResults, ". ")
 		cond := createValidationFailedCondition(msg)
 		r.result.Requeue = false
 		return buildStatusUpdateStateFnWithCondition(cond), nil
 	}
 	return stateFnInitialize, nil
 }
 
-func validateFunctionResources(rc *serverlessv1alpha2.ResourceConfiguration, minCPU resource.Quantity, minMemory resource.Quantity) []string {
-	if rc != nil && rc.Function != nil && rc.Function.Resources != nil {
-		vrLimits := validateLimits(*rc.Function.Resources, minMemory, minCPU, "Function")
-		vrRequests := validateRequests(*rc.Function.Resources, minMemory, minCPU, "Function")
-		vr := append(vrLimits, vrRequests...)
-		return vr
+func validateFunctionResourcesFn(rc *serverlessv1alpha2.ResourceConfiguration, minCPU resource.Quantity, minMemory resource.Quantity) validationFn {
+	return func() []string {
+		if rc != nil && rc.Function != nil && rc.Function.Resources != nil {
+			vrLimits := validateLimits(*rc.Function.Resources, minMemory, minCPU, "Function")
+			vrRequests := validateRequests(*rc.Function.Resources, minMemory, minCPU, "Function")
+			vr := append(vrLimits, vrRequests...)
+			return vr
+		}
+		return []string{}
+	}
+}
+
+func validateBuildResourcesFn(rc *serverlessv1alpha2.ResourceConfiguration, minCPU resource.Quantity, minMemory resource.Quantity) validationFn {
+	return func() []string {
+		if rc != nil && rc.Build != nil && rc.Build.Resources != nil {
+			vrLimits := validateLimits(*rc.Build.Resources, minMemory, minCPU, "Function")
```

Comment:
Do you thing about "Function" or maybe "Build"?

Reply:
Good catch, will fix it